File: MLP/read_tep.py
import numpy as np
from pathlib import Path

from mutual_information import mutual_information_matrix

class SIS_TEP:
    """
    A class for handling Temporal Event Process (TEP) data from SIS simulations.

    This class provides functionality to:
    - Load and process TEP data from various network types (e.g., erdos-renyi, barabasi-albert)
    - Sample system states at specified time points or intervals
    - Calculate and store mutual information matrices
    - Manage associated graph data

    Attributes:
        root (str): Root directory for specific network type and size
        network_type (str): Type of network (abbreviated)
        graph_id (str): Graph identifier
        tep_id (str): TEP identifier
        data (ndarray): Loaded TEP data
        time_points (ndarray): Transition time points
        vertex_indices (ndarray): Vertex indices for transitions
        N_vertices (int): Number of vertices
        max_T (float): Maximum time point

    Example:
        >>> tep = SIS_TEP("er", 100, 1, 1)
        >>> state = tep(0.5)  # Get system state at t=0.5
        >>> samples = tep.sample(0.1)  # Sample with dt=0.1
        >>> M = tep.generate_mutual_info(0.1)  # Generate mutual information matrix
        >>> M = tep.load_or_generate_mutual_info(0.1)  # Load mutual information matrix
    """
    
    __base_dir__ = "/home/DATA/datasets/TEP/gillespie_SIS/results/sis"
    __abrv_to_full__ = {
        "er": "erdos-renyi",
        "ba": "barabasi-albert",
        "ws": "watts-strogatz",
        "euc": "euclidean",
        "geo": "geometric",
        "grid": "grid",
        "reg": "regular",
        "sf": "scale-free",
    }

    def __init__(self, abrv, nb_vertices, i_graph, j_tep, nb_digits_g=2, nb_digits_tep=3):
        """
        Initialize the TEP reader.

        Args:
            abrv (str): Abbreviation for the network type.
            nb_vertices (int): Number of vertices.
            i_graph (int): Graph identifier.
            j_tep (int): TEP identifier.
            nb_digits_g (int, optional): Number of digits that the graph ID is stored in. Defaults to 2.
            nb_digits_tep (int, optional): Number of digits for TEP ID is stored in. Defaults to 3
        """
        self.root = f"{self.__base_dir__}/{self.__abrv_to_full__[abrv]}/N{nb_vertices}"
        self.network_type = abrv
        self.graph_id = f"{i_graph:0{nb_digits_g}d}"
        self.tep_id = f"{j_tep:0{nb_digits_tep}d}"
        self.load_tep()

    def get_tep_location(self):
        return f"{self.root}/tep-{self.network_type}-{self.graph_id}-{self.tep_id}.npz"

    def get_graph_location(self):
        return f"{self.root}/{self.network_type}-{self.graph_id}.npz"

    def get_mutual_info_location(self, dt):
        return f"{self.root}/mim-{self.network_type}-{self.graph_id}-{self.tep_id}-{dt:.2f}.npz"

    def get_sample_location(self, dt):
        return f"{self.root}/tep-{self.network_type}-{self.graph_id}-{self.tep_id}-{dt:.2f}.npy"

    def get_sample_location(self, dt, p):
        return f"{self.root}/tep-{self.network_type}-{self.graph_id}-{self.tep_id}-{dt:.2f}-p{p:.3f}.npy"

    def load_tep(self):
        """
        Load the TEP data, on the location specified by the constructor data.
        """
        filename = self.get_tep_location()
        self.data = np.load(filename)
        # time points at which a transition occurs
        self.time_points = self.data[:, 0]
        # vertex indices of the transitions
        self.vertex_indices = self.data[:, 1].astype(int) - 1
        self.N_vertices = max(self.vertex_indices) + 1
        self.max_T = self.time_points[-1]

    def __call__(self, t):
        """
        Returns the state of the system at time t.
        """
        if t < 0:
            raise ValueError("t must be non-negative.")
        idx = np.searchsorted(self.time_points, t, side='right')
        # The state of the system at time t is the parity of the number of times each vertex has transitioned up to time t.
        return np.array([np.count_nonzero(self.vertex_indices[:idx+1] == v) % 2 for v in range(self.N_vertices)])

    def __call__(self, t, p):
        """
        Returns the state of the system at time t with noise.
        The state of each vertex at time t is changed with probability p.
        """
        x = self(t)
        for i in range(self.N_vertices):
            if np.random.rand() < p:
                x[i] = 1 - x[i]

    def sample_at_ts(self, ts, p=0):
        """
        Sample with noise at time points ts.
        The state of each vertex at each time point is chosen randomly with probability p.
        """
        call_f = (lambda t: self(t, p)) if p > 0 else (lambda t: self(t))
        return np.array([call_f(t) for t in ts])

    def sample_with_dt(self, dt, p=0):
        """
        Samples the TEP with a time step dt and noise.
        """
        ts = np.arange(0, self.max_T + dt, dt)
        return self.sample_at_ts(ts, p) if p > 0 else self(ts)

    def sample(self, t, p=0):
        """
        Generate a TEP at time points ts (if ts is an array) or with a time step dt (if t is a number).
        If p is provided it acts as a probability of flipping the state of each vertex.
        """
        if isinstance(t, float):
            return self.sample_with_dt(t, p) if p > 0 else self.sample_with_dt(t)
        else:
            return self.sample_at_ts(t, p) if p > 0 else self.sample_at_ts(t)

    def store_sample(self, dt):
        """
        Store the TEP sample with time step dt at location
        "../tep-{network_type}-{graph_id}-{tep_id}-{dt:.2f}.npy".
        """
        np.save(self.get_sample_location(dt), self.sample_with_dt(dt))

    def store_sample(self, dt, p):
        """
        Store the TEP sample with time step dt and noise paramater p at location
        "../tep-{network_type}-{graph_id}-{tep_id}-{dt:.2f}-p{p:.3f}.npy".
        """
        np.save(self.get_sample_location(dt, p), self.sample_with_dt(dt, p))

    def load_graph(self):
        """
        Return the adjacency matrix of the graph associated with the TEP.
        """
        return np.load(self.get_graph_location())

    def generate_mutual_info(self, dt, store=True):
        """
        Generate the mutual information matrix for the TEP.
        """
        M = mutual_information_matrix(self.sample_with_dt(dt))
        if store:
            np.savez(self.get_mutual_info_location(dt), M=M)
        return M

    def load_mutual_info(self, dt):
        """
        Load the mutual information matrix for the TEP.
        """
        return np.load(self.get_mutual_info_location(dt))['M']

    def load_or_generate_mutual_info(self, dt):
        """
        Load the mutual information matrix if it exists, otherwise generate (and store) it.
        """
        path = Path(self.get_mutual_info_location(dt))
        if path.exists():
            try:
                return np.load(path)['M']
            except:
                pass
        return self.generate_mutual_info(dt)

File: gillespie_SIS/src/read_tep.jl
struct TEPLocation
    root::String
    network_type::String
    graph_id::String
    tep_id::String
end

struct TEP
    location::TEPLocation
    time_points::Vector{Float64}
    vertex_indices::Vector{Int}
    N_vertices::Int
    max_T::Float64
end

const ABRV_TO_FULL = Dict(
    "er" => "erdos-renyi",
    "ba" => "barabasi-albert",
    "ws" => "watts-strogatz",
    "euc" => "euclidean",
    "geo" => "geometric",
    "grid" => "grid",
    "reg" => "regular",
    "sf"  => "scale-free"
)

function abrv_to_full(abrv::String)::String
    if endswith(abrv, "md")
        return ABRV_TO_FULL[substring(abrv, 1, end-2)] * "-multi-degree"
    else
        return ABRV_TO_FULL[abrv]
    end
end

function load_tep(loc::TEPLocation)
    filename = get_tep_location(loc)
    data = npzread(filename)
    time_points = Vector{Float64}(data[:, 1])
    vertex_indices = Vector{Int}(Int.(data[:, 2]))
    N_vertices = maximum(vertex_indices)
    max_T = time_points[end]
    return TEP(loc, time_points, vertex_indices, N_vertices, max_T)
end

"""
    TEP(abrv::String, nb_vertices::Int, i_graph::Int, j_tep; nb_digits_g=2, nb_digits_tep=3)

Load the TEP with the given parameters.

# Arguments
- `abrv::String`: The abbreviation of the network type.
- `nb_vertices::Int`: The number of vertices in the network.
- `i_graph::Int`: The index of the graph.
- `j_tep::Int`: The index of the TEP.
- `nb_digits_g::Int=2`: The number of digits in the graph index.
- `nb_digits_tep::Int=3`: The number of digits in the TEP index.
"""
function TEP(abrv::String, nb_vertices::Int, i_graph::Int, j_tep::Int;
               nb_digits_g::Int=2, nb_digits_tep::Int=3)::TEP
    root = @sprintf("%s/%s/N%d", "/home/tim/Documents/overleaf/TEP/gillespie_SIS/results/sis",
                   abrv_to_full(abrv), nb_vertices)
    network_type = abrv
    graph_id = @sprintf("%0*d", nb_digits_g, i_graph)
    tep_id = @sprintf("%0*d", nb_digits_tep, j_tep)
    location = TEPLocation(root, network_type, graph_id, tep_id)
    return load_tep(location)
end

function get_tep_location(obj::TEPLocation)::String
    return "$(obj.root)/tep-$(obj.network_type)-$(obj.graph_id)-$(obj.tep_id).npz"
end
get_tep_location(obj::TEP)::String = get_tep_location(obj.location)

function get_graph_location(obj::TEPLocation)::String
    return "$(obj.root)/$(obj.network_type)-$(obj.graph_id).npz"
end
get_graph_location(obj::TEP)::String = get_graph_location(obj.location)

function get_sample_location(obj::TEPLocation, dt::Float64)::String
    return "$(obj.root)/tep-$(obj.network_type)-$(obj.graph_id)-$(obj.tep_id)-$(@sprintf("%.2f", dt)).npy"
end
get_sample_location(obj::TEP, dt::Float64)::String = get_sample_location(obj.location, dt)

function get_sample_location(obj::TEPLocation, dt::Float64, p::Float64)::String
    return "$(obj.root)/tep-$(obj.network_type)-$(obj.graph_id)-$(obj.tep_id)-$(@sprintf("%.2f", dt))-p$(@sprintf("%.3f", p)).npy"
end
get_sample_location(obj::TEP, dt::Float64, p::Float64)::String = get_sample_location(obj.location, dt, p)

function (obj::TEP)(t::Float64, p::Float64=0.0)::Vector{Int}
    if t < 0
        error("t must be non-negative.")
    end
    idx = searchsortedlast(obj.time_points, t)
    x = [count(x -> x == v, obj.vertex_indices[1:idx]) % 2 for v in 1:obj.N_vertices]

    if p > 0.0
        for i in 1:obj.N_vertices
            if rand() < p
                x[i] = 1 - x[i]
            end
        end
    end
    return x
end

function sample_at_ts(obj::TEP, ts::AbstractArray, p::Float64=0.0)
    return hcat([obj(t, p) for t in ts]...)'
end

function sample_with_dt(obj::TEP, dt::Float64, p::Float64=0.0)::Array{Int,2}
    ts = 0:dt:obj.max_T
    return sample_at_ts(obj, ts, p)
end

function sample(obj::TEP, t::Union{Float64, Vector{Float64}}, p::Float64=0.0)
    if isa(t, Float64)
        return sample_with_dt(obj, t, p)
    elseif isa(t, Vector{Float64})
        return sample_at_ts(obj, t, p)
    else
        error("Invalid type for t. Must be Float64 or Vector{Float64}.")
    end
end

function store_sample(obj::TEP, dt::Float64)
    data = sample_with_dt(obj, dt)
    filename = get_sample_location(obj, dt)
    Numpy.save(filename, data)
end

function store_sample_p(obj::TEP, dt::Float64, p::Float64)
    data = sample_with_dt(obj, dt, p)
    filename = get_sample_location(obj, dt, p)
    Numpy.save(filename, data)
end

function load_graph(obj::TEP)::Array{Float64,2}
    return npzread(get_graph_location(obj))
end

